tests/test_ingest.py

```diff
@@ -25,38 +25,336 @@
+        self.bad_metadata_file = os.path.join(TESTDIR, "data", "small.fits")
+        self.good_file = os.path.join(INGESTDIR, "sidecar_data", "dataset_2.yaml")
+        self.bad_instrument_file = os.path.join(TESTDIR, "data", "calexp.fits")
+
+    def testSimpleIngest(self):
+        # Use the default per-instrument run for this
+        self.task.run([self.good_file])
+        datasets = list(self.butler.registry.queryDatasets("raw_dict", collections="DummyCam/raw/all"))
+        self.assertEqual(len(datasets), 1)
+
+        # Now parallelized
+        files = [self.good_file,
+                 os.path.join(INGESTDIR, "sidecar_data", "dataset_1.yaml")]
+        self.task.run(files, processes=2, run=self.outputRun)
+        datasets = list(self.butler.registry.queryDatasets("raw_dict", collections=self.outputRun))
+        self.assertEqual(len(datasets), 2)
+
+    def testExplicitIndex(self):
+        files = [os.path.join(INGESTDIR, "indexed_data", "_index.json")]
+        self.task.run(files, run=self.outputRun)
+
+        datasets = list(self.butler.registry.queryDatasets("raw_dict", collections=self.outputRun))
+        self.assertEqual(len(datasets), 2)
+
+        # Try again with an explicit index and a file that is in that index
+        files.append(os.path.join(INGESTDIR, "indexed_data", "dataset_2.yaml"))
+        new_run = self.outputRun + "b"
+        self.task.run(files, run=new_run)
+
+        datasets = list(self.butler.registry.queryDatasets("raw_dict", collections=self.outputRun))
+        self.assertEqual(len(datasets), 2)
+
+        # Now with two index files that point to the same files.
+        # Look for the warning from duplication.
+        files = [os.path.join(INGESTDIR, "indexed_data", "_index.json"),
+                 os.path.join(INGESTDIR, "indexed_data", "translated_subdir", "_index.json")]
+        new_run = self.outputRun + "c"
+
+        with self.assertLogs(level="WARNING") as cm:
+            with lsst.log.UsePythonLogging():
+                self.task.run(files, run=new_run)
+        self.assertIn("already specified in an index file, ignoring content", cm.output[0])
+
+        datasets = list(self.butler.registry.queryDatasets("raw_dict", collections=self.outputRun))
+        self.assertEqual(len(datasets), 2)
+
+        # Again with an index file of metadata and one of translated.
+        # Translated should win.
+        # Put the metadata one first to test that order is preserved.
+        files = [os.path.join(INGESTDIR, "indexed_data", "metadata_subdir", "_index.json"),
+                 os.path.join(INGESTDIR, "indexed_data", "_index.json")]
+        new_run = self.outputRun + "d"
+        with self.assertLogs(level="WARNING") as cm:
+            with lsst.log.UsePythonLogging():
+                self.task.run(files, run=new_run)
+        self.assertIn("already specified in an index file but overriding", cm.output[0])
+
+        # Reversing the order should change the warning
```

Comment:
period at end of sentences

Reply:
I have to admit that I had no idea that periods were required at the end of short one-line comments. I've looked at the developer guide and it's not obvious but PEP-8 says "complete sentences" and does indeed have one example comment in there that ends in a period (but not for inline comments). For my entire career  I've always dropped the period for one-line comments and used the period for multi-line. The period seemed superfluous in a single line. Our code is full of cases where that is happening and you are the first person to mention it.